cedar-policy-validator/src/custom_schema/fmt.rs

```diff
@@ -0,0 +1,198 @@
+    SchemaTypeVariant, ValidatorSchema,
+};
+
+impl Display for SchemaFragment {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        for (ns, def) in &self.0 {
+            if ns.is_empty() {
+                write!(f, "{def}")?
+            } else {
+                write!(f, "namespace {ns} {{{def}}}")?
+            }
+        }
+        Ok(())
+    }
+}
+
+impl Display for NamespaceDefinition {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        for (n, ty) in &self.common_types {
+            writeln!(f, "type {n} = {ty};")?
+        }
+        for (n, ty) in &self.entity_types {
+            writeln!(f, "entity {n} {ty};")?
+        }
+        for (n, a) in &self.actions {
+            writeln!(f, "action \"{}\" {a};", n.escape_debug())?
+        }
+        Ok(())
+    }
+}
+
+impl Display for SchemaType {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            SchemaType::Type(ty) => match ty {
+                SchemaTypeVariant::Boolean => write!(f, "__cedar::Bool"),
+                SchemaTypeVariant::Entity { name } => write!(f, "{name}"),
+                SchemaTypeVariant::Extension { name } => write!(f, "__cedar::{name}"),
+                SchemaTypeVariant::Long => write!(f, "__cedar::Long"),
+                SchemaTypeVariant::Record {
+                    attributes,
+                    additional_attributes: _,
+                } => {
+                    write!(f, "{{ ")?;
+                    for (n, ty) in attributes {
+                        writeln!(
+                            f,
+                            "\"{}\"{}: {},",
+                            n.escape_debug(),
+                            if ty.required { "" } else { "?" },
+                            ty.ty
+                        )?;
+                    }
+                    write!(f, " }}")?;
+                    Ok(())
+                }
+                SchemaTypeVariant::Set { element } => write!(f, "Set < {element} >"),
+                SchemaTypeVariant::String => write!(f, "__cedar::String"),
+            },
+            SchemaType::TypeDef { type_name } => write!(f, "{type_name}"),
+        }
+    }
+}
+
+fn fmt_vec<T: Display>(f: &mut std::fmt::Formatter<'_>, ets: &[T]) -> std::fmt::Result {
+    match ets.split_last() {
+        Some((tail, head)) => {
+            if head.is_empty() {
+                write!(f, "{tail}")
+            } else {
+                write!(f, "[")?;
+                for et in head {
+                    write!(f, "{et}, ")?
+                }
+                write!(f, "{tail}")?;
+                write!(f, "]")?;
+                Ok(())
+            }
+        }
+        // PANIC SAFETY: input should be non-empty a slice
+        #[allow(clippy::unreachable)]
+        None => unreachable!("input list must not be empty"),
```

Comment:
This seems like an invariant that should be documented in this function's doc string, at a minimum.